loader/extends.go

```diff
@@ -35,96 +35,120 @@ func ApplyExtends(ctx context.Context, dict map[string]any, opts *Options, track
+		if !ok {
+			return nil, fmt.Errorf("cannot extend service %q in %s: service not found", name, filename)
 		}
+	}
+	// recursively apply `extends`
+	base, err = applyServiceExtends(ctx, ref, services, opts, tracker, post...)
+	if err != nil {
+		return nil, err
+	}
 
-		var base any
-		if file != nil {
-			path := file.(string)
-			for _, loader := range opts.ResourceLoaders {
-				if !loader.Accept(path) {
-					continue
-				}
-				local, err := loader.Load(ctx, path)
-				if err != nil {
-					return err
-				}
-				localdir := filepath.Dir(local)
-				relworkingdir := loader.Dir(path)
+	source := deepClone(base).(map[string]any)
+	for _, processor := range post {
+		processor.Apply(map[string]any{
+			"services": map[string]any{
+				name: source,
+			},
+		})
+	}
+	merged, err := override.ExtendService(source, service)
+	if err != nil {
+		return nil, err
+	}
+	delete(merged, "extends")
+	return merged, nil
+}
 
-				extendsOpts := opts.clone()
-				extendsOpts.ResourceLoaders = append([]ResourceLoader{}, opts.ResourceLoaders...)
-				// replace localResourceLoader with a new flavour, using extended file base path
-				extendsOpts.ResourceLoaders[len(opts.ResourceLoaders)-1] = localResourceLoader{
-					WorkingDir: localdir,
-				}
-				extendsOpts.ResolvePaths = true
-				extendsOpts.SkipNormalization = true
-				extendsOpts.SkipConsistencyCheck = true
-				extendsOpts.SkipInclude = true
-				source, err := loadYamlModel(ctx, types.ConfigDetails{
-					WorkingDir: relworkingdir,
-					ConfigFiles: []types.ConfigFile{
-						{Filename: local},
-					},
-				}, extendsOpts, ct, nil)
-				if err != nil {
-					return err
-				}
-				services := source["services"].(map[string]any)
-				base, ok = services[ref]
-				if !ok {
-					return fmt.Errorf("cannot extend service %q in %s: service not found", name, path)
-				}
-			}
-			if base == nil {
-				return fmt.Errorf("cannot read %s", path)
-			}
-		} else {
-			base, ok = services[ref]
-			if !ok {
-				return fmt.Errorf("cannot extend service %q in %s: service not found", name, "filename") // TODO track filename
-			}
+func getExtendsBaseFromFile(ctx context.Context, name string, path string, opts *Options, ct *cycleTracker) (map[string]any, error) {
+	for _, loader := range opts.ResourceLoaders {
+		if !loader.Accept(path) {
+			continue
 		}
-		source := deepClone(base).(map[string]any)
-		for _, processor := range post {
-			processor.Apply(map[string]any{
-				"services": map[string]any{
-					name: source,
-				},
-			})
+		local, err := loader.Load(ctx, path)
+		if err != nil {
+			return nil, err
+		}
+		localdir := filepath.Dir(local)
+		relworkingdir := loader.Dir(path)
+
+		extendsOpts := opts.clone()
+		extendsOpts.ResourceLoaders = append([]ResourceLoader{}, opts.ResourceLoaders...)
+		// replace localResourceLoader with a new flavour, using extended file base path
+		extendsOpts.ResourceLoaders[len(opts.ResourceLoaders)-1] = localResourceLoader{
+			WorkingDir: localdir,
 		}
```

Comment:
Instead of assuming that `localResourceLoader` is the last in the slice, it might be better to loop through and attempt to cast them until you find it and then replace it

Reply:
I introduced `RemoteResourceLoaders` to check this implicit rule